# The Battle of Neighborhoods - Final Assignment

In this Final Assignment we will examine the Neighborhoods of Brussels in order to identify the best Neighborhoods for a young couple that recently moved to the Belgian capital. The Real Estate company recently added a additional service that offers their clients a tailored service including a detailed report on the best Neighborhoods based on the preferences of the client.  

### Table of Content

* [1. Introduction: Business Problem](#Introduction)
* [2. Data](#Data)
    * [2.1 Scrapping and Cleaning](#Scrapping_and_Cleaning)
    * [2.2 Latitude and Longitude Data](#Latitude_and_Longitude)
    * [2.3 Visualize the map](#Visualize_Map)
    * [2.4 List of venues](#Foursquare_venues)
* [3. Methodology](#Methodology)

## 1. Introduction: Business Problem <a class="anchor" id="Introduction"></a>

A young couple recently moved to the Belgian capital for work and are wondering which of the Neighborhoods of Brussels will be the best to live in. Their interests include:  
1. The outdoors
2. Italian Restaurants
3. Cultural activities

They just saw an advertisment for a tailored service from a Real Estate company that will help them identify the top 3 Neighborhoods to choose based on their interests. Therefore, they contacted this Real Estate company and transmitted their interests.  
The young couple both are in their 30s and work full-time and don't know Brussels at all. As they will have to move in the next weeks to the Belgian capital they will need to find a new appartment quickly. 
The Real Estate company told them that they could find them an appartment in two weeks. The first week would be dedicated in coming up with the top 3 Neighbourhoods tailored to the couple and the second week would be used to find appartments in these Neighbourhoods.

The young couple are quite excited in seeing the analysis of the Real Estate and are curious what kind of Neighborhoods the Real Estate company will come up with. As they don't have any aquaintance in the city they depend 100% from the Real Estate company.

The Real Estate company analyzes their interests and determines that the best way to come up with the best Neighborhoods they will need the coordinates of Brussels and combine this with the Foursquare location data. 

## 2. Data <a class="anchor" id="Data"></a>

In order to determine the best Neighborhoods for the young couple the Real Estate company first analyzes how Brussels is divided.  
Brussels is the capital of Belgium and is divided in 19 municipalities, each whith a defined Postal Code.  
The municipalities can also be divided in different Neighborhoods. However, the Real Estate company finds out that these smaller sections don't have very defined limits and no specific Postal Codes.

Therefore, they decided to restrict their recommendation to one of the 19 municipalities of Brussels.  

Furthermore regarding these findings the Real Estate company decides that the data they will need is:

       - A list of each 19 municipalities of Brussels combined with their postal code and the geographic coordinates.
       - Categories of the venues inside these 19 municipalities. 

1. *List of 19 municipalities*  
For the list of each 19 municipalities of Brussels, the Real Estate company found a Wikipedia page that includes the 19 municipalities with their corresponding postal codes. They will scrap the Wikipedia page to extract the table.  
However, they still need to get the latitude and longitude coordinates for each postal code. For this they found a CSV file from the Open Data platform of Wallonia and Brussels that gives the full list of each postal code in Belgium combined with its respective longitude and latitude data.  

2. *Categories of venues*  
In order to get a list of venues and their categories the Real Estate company decides to use the Foursquare API and create a venue list based on the geographic coordinates identified previously.  

To find a list of the 19 Brussels Municipalities the Real Estate company found the Wikipedia page that has a table that lists the 19 municipalities with their corresponding Postal Codes.

### 2.1 Scrapping and Cleaning the Wikipedia page <a class="anchor" id="Scrapping_and_Cleaning"></a>

To start the company scrapes the Wikipedia page that hosts the Table of the different 19 municipalities of Brussels.  

The first step is to install and import the different modules we will need to scrape and clean the Table. For this operation we will need Pandas and Numpy.

In [1]:
import pandas as pd
import numpy as np

To scrape the Wikipedia page the company decides to use the Pandas read_html method that returns HTML tables into a list of DataFrame objects.  They prefer this method to the BeautifulSoup package as the pandas method is more straightforward for extracting table data from a HTML page.

The URL of the Wikipedia page that contains the Table is: "https://fr.wikipedia.org/wiki/R%C3%A9gion_de_Bruxelles-Capitale"

In [2]:
url = 'https://fr.wikipedia.org/wiki/R%C3%A9gion_de_Bruxelles-Capitale'

df_list = pd.read_html(url)

len(df_list)

16

We can see that read_html identified 16 different HTML tables. From navigating to the Wikipedia page we can identify that the table we are interested in is the second HTML table present. Therefore, let's only look at the second DataFrame object.

In [3]:
brussels_list = df_list[1]

brussels_list.head()

Repèresur lacarte                                            Commune  \
0                1.0                                         Anderlecht   
1                2.0                                          Auderghem   
2                3.0                              Berchem-Sainte-Agathe   
3                4.0  Ville de Bruxelles : Bruxelles Laeken Neder-Ov...   
4                5.0                                          Etterbeek   

                                   Code postal  Superficie(en km2)  \
0                                         1070                1774   
1                                         1160                 903   
2                                         1082                 295   
3  Plusieurs : 1000, 1040, 1050 1020 1120 1130                3261   
4                                         1040                 315   

  Population(au 1/1/2009) Population(au 1/1/2010) Population(au 1/1/2016)  \
0                 101 371                 104 647                 116 494   
1                  30 456                  30 811                  33 085   
2                  21 669                  22 185                  24 176   
3                 153 377                 157 673                 176 512   
4                  43 512                  44 352                  47 023   

  Population(au 1/1/2019)[36] Population(au 1/1/2020)[37] Évolution 2010-2020  \
0                     118 920                     120 009            + 15 362   
1                      33 970                      34 342             + 3 531   
2                      25 195                      25 195             + 3 010   
3                     179 797                     183 287            + 25 614   
4                      48 008                      48 194             + 3 842   

          Δ%  
0  + 14,68 %  
1  + 11,46 %  
2  + 13,56 %  
3  + 16,24 %  
4   + 8,66 %

From the above DataFrame we can see that the table includes a lot of information that we don't need for the moment. Let's only keep the relevant columns, "Commune" and "Code postal" which translates into "municipalities" and "Postal Code". 

In [4]:
brussels_list =  brussels_list [['Commune', 'Code postal']]

brussels_list

Commune  \
0                                          Anderlecht   
1                                           Auderghem   
2                               Berchem-Sainte-Agathe   
3   Ville de Bruxelles : Bruxelles Laeken Neder-Ov...   
4                                           Etterbeek   
5                                               Evere   
6                                              Forest   
7                                           Ganshoren   
8                                             Ixelles   
9                                               Jette   
10                                         Koekelberg   
11                               Molenbeek-Saint-Jean   
12                                       Saint-Gilles   
13                              Saint-Josse-ten-Noode   
14                                         Schaerbeek   
15                                              Uccle   
16                                Watermael-Boitsfort   
17                               Woluwe-Saint-Lambert   
18                                Woluwe-Saint-Pierre   
19                                             Totaux   

                                    Code postal  
0                                          1070  
1                                          1160  
2                                          1082  
3   Plusieurs : 1000, 1040, 1050 1020 1120 1130  
4                                          1040  
5                                          1140  
6                                          1190  
7                                          1083  
8                                          1050  
9                                          1090  
10                                         1081  
11                                         1080  
12                                         1060  
13                                         1210  
14                                         1030  
15                                         1180  
16                                         1170  
17                                         1200  
18                                         1150  
19                                          NaN

The table looks clean. However, we can see that the fourth row has more than one name and more than one postal code. That's because the central municipality of Brussels includes, aside of itself 3 other regions that formerly were seperate municipalities but are now attached to the municipality of Brussels. As they were formerly municipalities these 3 regions have distinct postal codes.  

Therefore, we will add these regions to the complete DataFrame.  

First let's have a look to the fourth row in detail.

In [5]:
city_brussels_PostalCode = brussels_list.iloc[3]

city_brussels_PostalCode

Commune        Ville de Bruxelles : Bruxelles Laeken Neder-Ov...
Code postal          Plusieurs : 1000, 1040, 1050 1020 1120 1130
Name: 3, dtype: object

In [6]:
city_brussels_name = city_brussels_PostalCode['Commune']

print(city_brussels_name)

Ville de Bruxelles : Bruxelles Laeken Neder-Over-Heembeek Haren


We can see that the row includes 4 distinct names. Namely: Bruxelles, Laeken, Neder-Over-Heembeek, and Haren. Therefore, let's create a list that includes these 4 municipalities names. 

In [7]:
city_brussels_name_list = city_brussels_name.split(" ")
city_brussels_name_list

['Ville',
 'de',
 'Bruxelles\xa0:',
 'Bruxelles',
 'Laeken',
 'Neder-Over-Heembeek',
 'Haren']

In [8]:
city_brussels_name_lists = city_brussels_name_list[3:7]

city_brussels_name_lists

['Bruxelles', 'Laeken', 'Neder-Over-Heembeek', 'Haren']

Let's have a look at the postal codes in this row. 

In [9]:
city_brussels_PostalCode_lists = city_brussels_PostalCode['Code postal'].split(" ")

city_brussels_PostalCode_lists

['Plusieurs\xa0:', '1000,', '1040,', '1050', '1020', '1120', '1130']

We can see that there are more postal codes than names. That's because "Bruxelles" had a sub-list of postal codes; 1000, 1040, and 1050. 

In [10]:
city_brussels_PostalCode_lists = city_brussels_PostalCode_lists[1:7]

city_brussels_PostalCode_lists

['1000,', '1040,', '1050', '1020', '1120', '1130']

Upon further inspection we can see that two of these postal codes; 1040 and 1050, are already present in the table. Therefore, we will only keep 1000 for Bruxelles.

In [11]:
del city_brussels_PostalCode_lists[1:3]

The last step is to remove the comma that appears after 1000.

In [12]:
city_brussels_PostalCode_lists = [city_brussels_PostalCode_lists[0].replace('1000,', '1000'), city_brussels_PostalCode_lists[1],city_brussels_PostalCode_lists[2], city_brussels_PostalCode_lists[3]]

city_brussels_PostalCode_lists

['1000', '1020', '1120', '1130']

With both list clean we are ready to merge them and create a sub-DataFrame.

In [13]:
brussels_city_df = pd.DataFrame({'Commune':city_brussels_name_lists, 'Code postal': city_brussels_PostalCode_lists})

brussels_city_df

Commune Code postal
0            Bruxelles        1000
1               Laeken        1020
2  Neder-Over-Heembeek        1120
3                Haren        1130

Now its time to fuse the two DataFrames together.

In [14]:
full_df = brussels_list.append(brussels_city_df, ignore_index=True)

full_df

Commune  \
0                                          Anderlecht   
1                                           Auderghem   
2                               Berchem-Sainte-Agathe   
3   Ville de Bruxelles : Bruxelles Laeken Neder-Ov...   
4                                           Etterbeek   
5                                               Evere   
6                                              Forest   
7                                           Ganshoren   
8                                             Ixelles   
9                                               Jette   
10                                         Koekelberg   
11                               Molenbeek-Saint-Jean   
12                                       Saint-Gilles   
13                              Saint-Josse-ten-Noode   
14                                         Schaerbeek   
15                                              Uccle   
16                                Watermael-Boitsfort   
17                               Woluwe-Saint-Lambert   
18                                Woluwe-Saint-Pierre   
19                                             Totaux   
20                                          Bruxelles   
21                                             Laeken   
22                                Neder-Over-Heembeek   
23                                              Haren   

                                    Code postal  
0                                          1070  
1                                          1160  
2                                          1082  
3   Plusieurs : 1000, 1040, 1050 1020 1120 1130  
4                                          1040  
5                                          1140  
6                                          1190  
7                                          1083  
8                                          1050  
9                                          1090  
10                                         1081  
11                                         1080  
12                                         1060  
13                                         1210  
14                                         1030  
15                                         1180  
16                                         1170  
17                                         1200  
18                                         1150  
19                                          NaN  
20                                         1000  
21                                         1020  
22                                         1120  
23                                         1130

We can see that this Dataframe still includes the 'faulty' row. Furthermore, it includes a Total row that has a *NaN* value. Therefore, let's drop these two rows.

In [15]:
full_df.drop([3, 19], inplace=True)

Lastly, let's sort the DataFrame by increasing number of postal code.

In [16]:
full_df.sort_values(by=['Code postal'], ignore_index=True)

Commune Code postal
0               Bruxelles        1000
1                  Laeken        1020
2              Schaerbeek        1030
3               Etterbeek        1040
4                 Ixelles        1050
5            Saint-Gilles        1060
6              Anderlecht        1070
7    Molenbeek-Saint-Jean        1080
8              Koekelberg        1081
9   Berchem-Sainte-Agathe        1082
10              Ganshoren        1083
11                  Jette        1090
12    Neder-Over-Heembeek        1120
13                  Haren        1130
14                  Evere        1140
15    Woluwe-Saint-Pierre        1150
16              Auderghem        1160
17    Watermael-Boitsfort        1170
18                  Uccle        1180
19                 Forest        1190
20   Woluwe-Saint-Lambert        1200
21  Saint-Josse-ten-Noode        1210

In [17]:
full_df.shape

(22, 2)

We see that the DataFrame now includes 21 postal codes, which corresponds to the 19 municipalities of Brussels plus the two added sub-regions of the Brussels municipality.

In order to get the longitutde and latitude coordinates of the 19 different municipalities, the Real Estate company has access to the Open Data website of Wallonia and Brussels that hosts a CSV file listing all Belgian Postal Codes with the names of the municipalities as well as their longitude and latitudes.


## 2.2 Adding the Latitude and Longitude data to the municipalities. <a class="anchor" id="Latitude_and_Longitude"></a>

In this section we will explore how to add the Latitude and Longitude coordinates to the municipalities using the postal codes as an index. 

The CSV used is openly accessible from the Open Data Wallonie-Bruxelles platform. [Link](https://www.odwb.be/explore/dataset/code-postaux-belge/information/)

As this CSV uses a semicolon to seperate the data we will have to specify this when reading the CSV into a DataFrame.

In [18]:
csv = 'https://www.odwb.be/explore/dataset/code-postaux-belge/download/?format=csv&timezone=Europe/Berlin&lang=fr&use_labels_for_header=true&csv_separator=%3B'

lat_long = pd.read_csv(csv, sep=';')

lat_long.head()

Code             Localite  Longitude   Latitude               Coordonnees  \
0  1050              Ixelles   4.381571  50.822285      50.8222854,4.3815707   
1  1081           Koekelberg   4.325708  50.862263      50.8622633,4.3257084   
2  1130                Haren   4.412571  50.891966  50.8919661,4.41257087324   
3  1150  Woluwe-Saint-Pierre   4.443297  50.829243      50.8292429,4.4432967   
4  1315              Incourt   4.798944  50.691800       50.6918004,4.798944   

   Geom  
0   NaN  
1   NaN  
2   NaN  
3   NaN  
4   NaN

Like before, we don't need all the columns present in the CSV. The columns that interest us are the Longitude and Latitude. Furthermore, we will need one column to act as a key when we will merge the two DataFrame together. As the postal code is unique the 'Code' column will also have to be kept.

In [19]:
lat_long = lat_long[['Code', 'Latitude', 'Longitude']]

Moreover, in order to use the postal code as an index we will have to rename the column to match the column of the postal code in our **brussels_df** DataFrame. 

In [20]:
lat_long.rename(columns = {'Code': 'Code postal'}, inplace = True)

We are ready to merge the **full_df** DataFrame with the **lat_long** DataFrame on the common column 'Code postal'. 
This will add the latitude and longitude coordinate to each municipality identified previously.

In [21]:
brussels_df = full_df.merge(lat_long,on=['Code postal'])

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

The error message indicated that the columns in the two DataFrames 'Code postal' used as a unique key don't have the same datatype. Let's have a look at both DataFrames and identify the data types of each column.

In [22]:
full_df.dtypes

Commune        object
Code postal    object
dtype: object

We can see that the two columns of the **full_df** DataFrame have the data type object. This is correct for the 'Commune' columns as we are speaking about the names of the municipalities. However, the 'Code postal' column should not have the object datatype but a numeric data type.  

Now let's look at the data types of the **lat_long** DataFrame.

In [23]:
lat_long.dtypes

Code postal      int64
Latitude       float64
Longitude      float64
dtype: object

We see that the column 'Code postal' of the **lat_long** DataFrame has the data type *int64*. Therefore, let's convert the 'Code postal' of the **full_df** DataFrame to int64.  
That way can then finally merge both DataFrames. 

In [24]:
full_df = full_df.astype({'Code postal':'int64'})

Let's verify if the data type has been well converted.

In [25]:
full_df.dtypes

Commune        object
Code postal     int64
dtype: object

We can see now that the 'Code postal' column in both DataFrames have the same data type and can be safely merged on this column.

In [26]:
brussels_df = full_df.merge(lat_long,on=['Code postal'])

In [27]:
brussels_df

Commune  Code postal   Latitude  Longitude
0              Anderlecht         1070  50.838141   4.312340
1               Auderghem         1160  50.815657   4.433139
2   Berchem-Sainte-Agathe         1082  50.863984   4.292702
3               Etterbeek         1040  50.836851   4.389510
4                   Evere         1140  50.870452   4.402160
5                  Forest         1190  50.809143   4.317751
6               Ganshoren         1083  50.871240   4.317510
7                 Ixelles         1050  50.822285   4.381571
8                   Jette         1090  50.877763   4.326090
9              Koekelberg         1081  50.862263   4.325708
10   Molenbeek-Saint-Jean         1080  50.854355   4.322778
11           Saint-Gilles         1060  50.826741   4.345668
12  Saint-Josse-ten-Noode         1210  50.853074   4.372336
13             Schaerbeek         1030  50.867604   4.373712
14                  Uccle         1180  50.801820   4.337235
15    Watermael-Boitsfort         1170  50.799394   4.415818
16   Woluwe-Saint-Lambert         1200  50.846693   4.428484
17    Woluwe-Saint-Pierre         1150  50.829243   4.443297
18              Bruxelles         1000  50.846557   4.351697
19                 Laeken         1020  50.883392   4.348713
20    Neder-Over-Heembeek         1120  50.897796   4.390489
21                  Haren         1130  50.891966   4.412571

The brussels_df DataFrame now includes the names of the municipalities, their corresponding postal code, and their latitude and longitude coordinates. This DataFrame can then be used to identify the municipalities on a map.

## 2.3 Visualization of the municipalities on a map <a class="anchor" id="Visualize_Map"></a>

In this section we will visualize the identified municipalities on a Map of Brussels.  

First lets install and import the need packages and modules.

In [28]:
import folium #package used to visualize a map

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import numpy as np # library to handle data in a vectorized manner

print('packages installed and imported')

import json

import requests

packages installed and imported


Let's use the Nominatim geolocator to find out the latitude and longitude of Brussels. 

In [29]:
address = 'Brussels'

geolocator = Nominatim(user_agent="br_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Brussels are: {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Brussels are: 50.8465573, 4.351697.


We'll use Brussels latitude and longtiude to center the Folium map and add markers to the identified municipalities on the map. 

In [30]:
map_brussels = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, municipality in zip(brussels_df['Latitude'], brussels_df['Longitude'], brussels_df['Commune']):
    label = '{}'.format(municipality)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_brussels)  
    
map_brussels

Once we have the latitude and longitude of the identified municipalities of Brussels we can then use Foursquare's location data to identify the type of venues in the different municipalities. The type of venues combined with the frequency of these venues inside of the identified municipalities could give us a good idea of which municipality would match the best the criteria's given by the young couple.

## 2.4 List of venues in the municipalities by Category <a class="anchor" id="Foursquare_venues"></a>

In this section we will make an API call to the Foursquare API in order to get information about the venues that are located within the municipalities we identifed earlier. For this we will create a function called getNearbyVenues and that will read into a new DataFrame the name of the venue combined with the municipality in which it is located and the coordinates of that municipality as well as the coordinates of the venues location and its category. 

First lets define the Foursquare API credentials to make it easier afterwards to recall the needed information.

In [31]:
CLIENT_ID = 'NIDDKXAR0Y1JRGEEWFFQHC234U455BN21PQDFE3EFDRP1GJC' # your Foursquare ID
CLIENT_SECRET = '2FO3JXOSKZJLNJKOYE3L120JFVE1T2S5UKDCJC5NGP2PKVST' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

Secondly, we will define the function called getNearbyVenues that will loop through the names of the municipalities and return data about the venues that are located within these municipalities. The radius was set at 500m and the API limit is set at 100. 

In [32]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now let's make use of this new function and read the **brussels_df** into it.

In [33]:
brussels_venues = getNearbyVenues(names=brussels_df['Commune'], latitudes = brussels_df['Latitude'], longitudes=brussels_df['Longitude'])

Anderlecht
Auderghem
Berchem-Sainte-Agathe
Etterbeek
Evere
Forest
Ganshoren
Ixelles
Jette
Koekelberg
Molenbeek-Saint-Jean
Saint-Gilles
Saint-Josse-ten-Noode
Schaerbeek
Uccle
Watermael-Boitsfort
Woluwe-Saint-Lambert
Woluwe-Saint-Pierre
Bruxelles
Laeken
Neder-Over-Heembeek
Haren


Let's have look at the first 5 rows of the venues identified thanks to the Foursquare API call. 

In [34]:
brussels_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0   Anderlecht              50.838141                 4.31234   
1   Anderlecht              50.838141                 4.31234   
2   Anderlecht              50.838141                 4.31234   
3   Anderlecht              50.838141                 4.31234   
4   Anderlecht              50.838141                 4.31234   

                                         Venue  Venue Latitude  \
0                                 Friture René       50.835846   
1                                 Snack Mirvan       50.835176   
2                                       Ulysse       50.838612   
3  Erasmushuis / Maison d'Erasme (Erasmushuis)       50.836507   
4                             Le Chapeau Blanc       50.835034   

   Venue Longitude      Venue Category  
0         4.311632  Belgian Restaurant  
1         4.308543         Snack Place  
2         4.306860    Greek Restaurant  
3         4.307860      History Museum  
4         4.307790          Restaurant

Furthermore, let's see how many venues were identified for all the municipalities of Brussels.

In [35]:
print(brussels_venues.shape)

(684, 7)


We see that we identified **684** different venues located in the municipalities of Brussels.  

In the next steps we will analyse these venues and identify the municipalities that include the most venues corresponding to the criterias stated by the young couple.

## 3. Methodology 

This section will use the Foursquare API to identify the different types of venues present in the different municipalities. Calculating the frequency of appearance of the different types of venues by municipalities will in turn be used to cluster the municipalities in similar municipalities.  
From there we wil analyze the different clusters in order to identify the clusters that respond the best to the criteria's given by the young couple. 